Updates to appveyor config for VS2019 and Qt5.9.8 + msvc project fixes

sipsorcery opened the pull request:
Updates to appveyor config:
 - Update build image from Visual Studio 2017 to Visual Studio 2019.
 - Updated Qt static library from Qt5.9.7 to Qt5.9.8.
 - Added commands to update vcpkg port files (this does not update already installed packages).
 - Updated vcpkg package list as per #17309.
 - Removed commands setting common project file options. Now done via common.init.vcxproj include.
 - Changed msbuild verbosity from normal to quiet. Normal writes a LOT of logs and impacts appveyor job duration.

Updates to msvc project configs:
 - Updated platform toolset from v141 to v142.
 - Updated Qt static library from Qt5.9.7 to Qt5.9.8.
 - Added ignore for linker warning building bitcoin-qt program.
 - Added missing util/str.cpp class file to test_bitcoin project file.

In order for an existing appveyor job based on the new config to work the cache must be purged. The steps to do this are shown below. The specific appveyor project path will need to be adjusted.

````
export APPVEYOR_TOKEN="<your-api-token>"
curl -H "Authorization: Bearer $APPVEYOR_TOKEN" -X DELETE https://ci.appveyor.com/api/projects/bitcoin/bitcoin-9ql6k/buildcache
````

laanwj commented:
Concept ACK

fanquake commented on .appveyor.yml:
```diff
@@ -44,27 +45,18 @@ before_build:
 - ps:  python test\util\bitcoin-util-test.py
 - cmd: python test\util\rpcauth-test.py
-- cmd: python test\functional\test_runner.py --ci --quiet --combinedlogslen=4000 --failfast
+# Fee estimation test fialing on appveyor with: WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted.
```
nit: s/fialing/failing

fanquake reviewed (commented):
Concept ACK - There's some related discussion in #17309 and these changes reflect that. i.e no longer needing OpenSSL as part of the Qt build and adding missing `vcpkg` packages.

Sjors commented:
Concept ACK. Works for me if it works for AppVeyor :-)

fanquake commented on .appveyor.yml:
```diff
@@ -1,31 +1,32 @@
 # Disable zmq test for now since python zmq library on Windows would cause Access violation sometimes.
 # - cmd: pip install zmq
 - cmd: echo set(VCPKG_BUILD_TYPE release) >> C:\tools\vcpkg\triplets\%PLATFORM%-windows-static.cmake
+- cmd: cd c:\tools\vcpkg
```
You mentioned [here](https://github.com/bitcoin/bitcoin/pull/17309#issuecomment-548083497) that updating `vcpkg` for each build would be infeasible. Is that no-longer the case, or am I mis-understanding what this does?

sipsorcery replied:
I checked my statement by running the appveyor job with a pull of the vcpkg repo as mentioned in the articles you [referenced](https://www.appveyor.com/docs/lang/cpp/#vc-packaging-tool) and I was incorrect about the time it would take. 

The reason being that updating and rebuilding the `vcpkg` program does not affect already installed packages. In my testing it only added about 2 minutes onto the build time. I wrongly assumed that updating vcpkg would invalidate the existing packages and force them to be re-installed.

For this PR the `vcpkg` source does have to be updated to get the latest port files otherwise the build fails (didn't track down why). After the first successful build the`vcpkg` update steps can be disabled to save the 2 minutes. It would need to be some kind of conditional disable so that anyone forking the repo still gets a working appveyor config. I can look into it if consensus is the small additional complexity justifies saving 2 minutes.

MarcoFalke commented:
This build took 1h14m. Other builds take typically between 33 min and 49 min. Are you sure caching has no effect on the run time?

sipsorcery commented:
> This build took 1h14m. Other builds take typically between 33 min and 49 min. Are you sure caching has no effect on the run time?

That does warrant some further investigation. I'll look into it further.

My build times performed sequentially were as below. They seemed fairly consistent even with appveyor's unpredictability. But given your results I'll see if maybe I missed something with clcache. It would certainly be nice to speed things up.

````
Disabled clcache
51 min 49 sec
53 min 46 sec
52 min 23 sec
53 min 16 sec
Failed after 47 min 38 sec
44 min 4 sec
44 min 2 sec

Enabled clcache
53 min 13 sec
52 min 26 sec
53 min 5 sec
54 min 45 sec
Failed after 46 min 26 sec
53 min 15 sec
52 min 34 sec
53 min 9 sec
53 min 54 sec

Disabled clcache
57 min 9 sec
Failed after 1 hr
59 min 25 sec
56 min 33 sec
54 min 55 sec
````

MarcoFalke commented:
Maybe clcache can be removed in a follow-up? Splitting the change into two commits would also help with review.

sipsorcery commented:
@MarcoFalke D'accord. Have put back the clcache steps. Now the changes in this PR are focused on getting the msvc build working with VS2019 and Qt5.9.8.

sipsorcery committed:
Updated appveyor config:
 - Update build image from Visual Studio 2017 to Visual Studio 2019.
 - Updated Qt static library from Qt5.9.7 to Qt5.9.8.
 - Added commands to update vcpkg port files (this does not update already installed packages).
 - Updated vcpkg package list as per #17309.
 - Removed commands setting common project file options. Now done via common.init.vcxproj include.
 - Changed msbuild verbosity from normal to quiet. Normal rights a LOT of logs and impacts appveyor job duration.
Updated msvc project configs:
 - Updated platform toolset from v141 to v142.
 - Updated Qt static library from Qt5.9.7 to Qt5.9.8.
 - Added ignore for linker warning building bitcoin-qt program.
 - Added missing util/str.cpp class file to test_bitcoin project file.

sipsorcery commented:
@MarcoFalke the appveyor job will need to have the cache cleared before this PR will build successfully. The Curl command below shows how to do it. You'll need to replace the project URL.

All the appveyor builds are failing at the moment due to a new utility class being added to the `test_bitcoin` program. This PR fixes that.

````
export APPVEYOR_TOKEN="<your-api-token>"
curl -H "Authorization: Bearer $APPVEYOR_TOKEN" -X DELETE https://ci.appveyor.com/api/projects/sipsorcery/bitcoin-72c17/buildcache
````

ryanofsky reviewed (approved):
Non-expert code review ACK 3c84deebaa311155ccc7565d09525041eed9747c.

Options changes look as expected and few lines of shell code added seem to do what's described in comments. I haven't looked into details of how appveyor build works, so I probably wouldn't notice any bugs, but the configuration looks reasonable and is very well commented

> @MarcoFalke the appveyor job will need to have the cache cleared before this PR will build successfully. The Curl command below shows how to do it. You'll need to replace the project URL.

Maybe we should have some permanent documentation that says how and when to run this command, if we don't already

MarcoFalke commented on .appveyor.yml:
```diff
@@ -1,31 +1,56 @@
+  VCPKG_INSTALL_PATH: 'C:\tools\vcpkg\installed'
 cache:
 - C:\tools\vcpkg\installed
 - C:\Users\appveyor\clcache -> .appveyor.yml, build_msvc\**, **\Makefile.am, **\*.vcxproj.in
```
Why is this cache not cleared? 

https://www.appveyor.com/docs/build-cache/#cleaning-up-cache

MarcoFalke commented:
```
$ curl -H "Authorization: Bearer $APPVEYOR_TOKEN" -X DELETE https://ci.appveyor.com/api/projects/bitcoin/bitcoin-9ql6k/buildcache
{"message":"Account not found or access denied."}

sipsorcery commented:
The cache purge worked. The appveyor job is now installing the vcpkg's from scratch. The first job will probably fail due to hitting the 1hr time limit. Installing the vcpkg's (~20mins) + build (~40mins) + test (~15mins) is now too long to do all at once. After the first attempt the vcpkg packages should be cached and subsequent job runs should work.

MarcoFalke commented:
Ok. Could you please update the OP to reflect the current state of this pull request?

sipsorcery commented:
OP and title updated.

MarcoFalke commented:
Just wanted to note that this reverts 0646ca5ea2541d4da324d9aff7de422c6daa0ef9, which needs to be applied manually on failing builds

fanquake commented:
post-merge ACK 3c84deebaa311155ccc7565d09525041eed9747c

I've just completed a qt build on Windows 10 with these changes:

![windows_17364](https://user-images.githubusercontent.com/863730/68403794-842df200-014b-11ea-9831-02e307397833.png)

Sjors commented:
The commit message was incorrectly formatted:

```
commit 3c84deebaa311155ccc7565d09525041eed9747c
Author: Aaron Clauson <aaron@sipsorcery.com>
Date:   Wed Nov 6 22:02:16 2019 +0000

    Updated appveyor config:
     - Update build image from Visual Studio 2017 to Visual Studio 2019.
```

There should be a blank line after the description. Otherwise things like `git rebase` get confused (though not broken), e.g.:

```
pick 037f2405f2 Updated appveyor config:  - Update build image from Visual Studio 2017 to Visual Studio 2019.  - Updated Qt static library from Qt5.9.7 to Qt5.9.8.  - Added commands to update vcpkg port files (this does not update already installed packages).  - Updated vcpkg package list as per #17309.  - Removed commands setting common project file options. Now done via common.init.vcxproj include.  - Changed msbuild verbosity from normal to quiet. Normal rights a LOT of logs and impacts appveyor job duration. Updated msvc project configs:  - Updated platform toolset from v141 to v142.  - Updated Qt static library from Qt5.9.7 to Qt5.9.8.  - Added ignore for linker warning building bitcoin-qt program.  - Added missing util/str.cpp class file to test_bitcoin project file.
pick 8a5d0954a3 configure: add ax_boost_process
pick 25405cc429 [build] msvc: add boost::process
```

MarcoFalke commented:
A linter makes sense to catch these

MarcoFalke commented:
It looks like compile time is back at about 50 minutes (like before): https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/28694137

Ok, this one is 57 minutes, but close enough.
